### ISSUE TITLE
404 Error transactions not marked as Error when displayed in dynatrace

### ISSUE THREAD

**Issue by harishsubramanian2577** [avioconsulting/mule-opentelemetry-module#87]
Hello Manik and team,
I am testing the agent for my mule proxy app downloaded from anypoint runtime manager. when I tested for 404 scenarios, the transaction is not marked as error and displays in dynatrace as if the transaction went through fine. Following are the scenarios i tested with the behaviors
1) 404 caught by the Rest valiadator module for a non existent resource in the API - for Ex:- /helloworld1 instead of actual resource /helloword. 
Behavior observed - In this case, the transaction is not showing up as an errored out transaction in "ingested trace" tab in Dynatrace, although the response code inside the transaction shows 404.
2) 404 error thrown by the back end API -  
Behavior observed - same behavior as scenario1, except that in this case, the http request checkpoint in the trace is marked as error with 404.
Attaching images. I can explain this better in a call.
<img width="1741" alt="404ErrorsNotMarkedAsErrorTxn" src="https://user-images.githubusercontent.com/116039895/201255068-b17e2cb5-b61a-4b2f-a4a5-1a67b11245b6.png">
<img width="1051" alt="Proxy404fromBackEnd" src="https://user-images.githubusercontent.com/116039895/201255076-a393fbe1-d297-4ffc-8396-edfe9f2f6aec.png">
<img width="1130" alt="Proxy404ResourceNotFound" src="https://user-images.githubusercontent.com/116039895/201255088-b5c755a3-61bb-41d7-a63a-ab51662e45fd.png">

Regards
Harish.S

**Comment by manikmagar** [comment]
@harishsubramanian2577 Exception event capturing to the spans is the same for any processor that generates Mule Errors. If an error exists in the mule-generated notification event, it will be captured on the spans. Are you by any chance using on-error-continue that may be removing the error from the event? Is it possible to add an example XML config here?

**Comment by harishsubramanian2577** [comment]
@manikmagar , I had already converted all the on-error-continue to on-error-propogate hoping it might fix the issue, but even after using on-error-propogate, the issue persists.

pasting the code below.`<?xml version="1.0" encoding="UTF-8"?>
<mule xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
      xmlns="http://www.mulesoft.org/schema/mule/core"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xmlns:http="http://www.mulesoft.org/schema/mule/http"
      xmlns:proxy="http://www.mulesoft.org/schema/mule/proxy"
      xmlns:ee="http://www.mulesoft.org/schema/mule/ee/core"
      xmlns:api-gateway="http://www.mulesoft.org/schema/mule/api-gateway"
      xmlns:opentelemetry="http://www.mulesoft.org/schema/mule/opentelemetry"
	  xmlns:otel-mule4-observablity-agent="http://www.mulesoft.org/schema/mule/otel-mule4-observablity-agent"
      xmlns:rest-validator="http://www.mulesoft.org/schema/mule/rest-validator" xsi:schemaLocation="http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
        http://www.mulesoft.org/schema/mule/http http://www.mulesoft.org/schema/mule/http/current/mule-http.xsd
        http://www.mulesoft.org/schema/mule/api-gateway http://www.mulesoft.org/schema/mule/api-gateway/current/mule-api-gateway.xsd
        http://www.mulesoft.org/schema/mule/proxy http://www.mulesoft.org/schema/mule/proxy/current/mule-proxy.xsd
        http://www.mulesoft.org/schema/mule/rest-validator http://www.mulesoft.org/schema/mule/rest-validator/current/mule-rest-validator.xsd
        http://www.mulesoft.org/schema/mule/ee/core http://www.mulesoft.org/schema/mule/ee/core/current/mule-ee.xsd
        http://www.mulesoft.org/schema/mule/otel-mule4-observablity-agent http://www.mulesoft.org/schema/mule/otel-mule4-observablity-agent/current/mule-otel-mule4-observablity-agent.xsd
		http://www.mulesoft.org/schema/mule/opentelemetry http://www.mulesoft.org/schema/mule/opentelemetry/current/mule-opentelemetry.xsd">

    <configuration-properties file="config.properties"/>

    <configuration defaultResponseTimeout="${proxy.responseTimeout}"/>

    <proxy:config name="proxy-config"/>
    
   
	
	<opentelemetry:config name="OpenTelemetry_Config" doc:name="OpenTelemetry Config" doc:id="8bf6f443-0d82-4f9b-8e25-1ca15d3b8f62" serviceName="api-app-1" spanAllProcessors="true">
		<opentelemetry:exporter >
			<opentelemetry:otlp-exporter collectorEndpoint="https://XXXXXX.live.dynatrace.com/api/v2/otlp/v1" protocol="HTTP_PROTOBUF" >
 				<opentelemetry:headers >
					<opentelemetry:header key="testHeader" value="testHeaderValue" />
					<opentelemetry:header key="Authorization" value="Api-Token XXXXX" />
				</opentelemetry:headers>
			</opentelemetry:otlp-exporter>
		</opentelemetry:exporter>



	</opentelemetry:config>


    <rest-validator:config name="configValidator"
                           api="${implementation.api.spec}"
                           parser="${implementation.api.parser}"
                           disableValidations="${validation.disable}"
                           queryParamsStrictValidation="${validation.strict.queryParams}"
                           headersStrictValidation="${validation.strict.headers}"/> 

    <flow name="proxy">
        <http:listener config-ref="http-listener-config" path="${proxy.path}">
            <http:response statusCode="#[attributes.statusCode]" reasonPhrase="#[attributes.reasonPhrase]">
                <http:headers><![CDATA[#[if (vars.proxyResponseHeaders?) vars.proxyResponseHeaders else {}]]]></http:headers>
            </http:response>
			<http:error-response statusCode="#[attributes.statusCode]" reasonPhrase="#[attributes.reasonPhrase]">
				<http:body ><![CDATA[#[payload]]]></http:body>
			</http:error-response>
        </http:listener>
        
        
        <rest-validator:validate-request config-ref="configValidator"/>

        <proxy:request-headers config-ref="proxy-config" target="proxyRequestHeaders"/>
        

        

        <http:request config-ref="http-request-config" method="#[attributes.method]" path="#[attributes.maskedRequestPath]" doc:id="be1bb696-764d-400b-b1cd-726ca4245600">
            <http:headers><![CDATA[#[vars.proxyRequestHeaders]]]></http:headers>
            <http:uri-params><![CDATA[#[attributes.uriParams]]]></http:uri-params>
            <http:query-params><![CDATA[#[attributes.queryParams]]]></http:query-params>
            <http:response-validator>
                <http:success-status-code-validator values="0..599"/>
            </http:response-validator>
        </http:request>

        <proxy:response-headers config-ref="proxy-config" target="proxyResponseHeaders"/>

        <logger level="INFO" doc:name="Logger" doc:id="b90a1d91-ee42-4bb8-bc67-04a0b0ed1f5e" message="responseHeaders:#[vars.proxyResponseHeaders]"/>
		<error-handler>
        
        	<on-error-propagate enableNotifications="true" logException="true" doc:name="On Error Propagate" doc:id="3973f9ec-dafa-4f43-a441-010c90262fa6" type="HTTP:TIMEOUT">
        		<logger level="INFO" doc:name="Logger" doc:id="3c560a27-5e36-417a-a956-875c3da65e74" message="Error Occurred: statusCode: 504"/>
                <ee:transform>
                    <ee:message>
                        <ee:set-payload>output application/json --- {"error": "504 Gateway timeout error"}</ee:set-payload>
                        <ee:set-attributes>
                            {
                            statusCode: 504
                            }
                        </ee:set-attributes>
                    </ee:message>
                    <ee:variables>
                        <ee:set-variable variableName="httpStatus"><![CDATA[504]]></ee:set-variable>
                    </ee:variables>
                </ee:transform>
			</on-error-propagate>
			
			<on-error-propagate enableNotifications="true" logException="true" doc:name="On Error Propagate" doc:id="f4a4bdbd-f33a-4fac-82fb-c80e8e8d30aa" type="REST-VALIDATOR:BAD_REQUEST">
        		<logger level="INFO" doc:name="Logger" doc:id="89524483-2690-443e-9999-3a7381e74064" message="Error Occurred: statusCode: 400"/>
                <ee:transform>
                    <ee:message>
                        <ee:set-payload>output application/json --- {"error": "$(error.description)"}</ee:set-payload>
                        <ee:set-attributes>
                            {
                            statusCode: 400
                            }
                        </ee:set-attributes>
                    </ee:message>
                    <ee:variables>
                        <ee:set-variable variableName="httpStatus"><![CDATA[400]]></ee:set-variable>
                    </ee:variables>
                </ee:transform>
			</on-error-propagate>
			
			<on-error-propagate enableNotifications="true" logException="true" doc:name="On Error Propagate" doc:id="3f885b43-4d25-4dec-bb13-5555fd5b6b46" type="REST-VALIDATOR:METHOD_NOT_ALLOWED">
        		<logger level="INFO" doc:name="Logger" doc:id="994dcb98-91a4-4827-9957-395ef1cbbff8" message="Error Occurred: statusCode: 405"/>
                <ee:transform>
                    <ee:message>
                        <ee:set-payload>output application/json --- {"error": "$(error.description)"}</ee:set-payload>
                        <ee:set-attributes>
                            {
                            statusCode: 405
                            }
                        </ee:set-attributes>
                    </ee:message>
                    <ee:variables>
                        <ee:set-variable variableName="httpStatus"><![CDATA[405]]></ee:set-variable>
                    </ee:variables>
                </ee:transform>
			</on-error-propagate>
            
            <on-error-propagate enableNotifications="true" logException="true" doc:name="On Error Propagate" doc:id="99836520-95aa-4fb0-aa96-b7048239b7cd" type="HTTP:NOT_FOUND, REST-VALIDATOR:RESOURCE_NOT_FOUND">
             <logger level="INFO" doc:name="Logger" doc:id="5e60990d-5ef2-48a9-82a5-d406fc054008" message="Error Occurred: statusCode: 404"/>
				<ee:transform doc:name="Transform Message" doc:id="a38af0d7-6cd8-4d49-843e-1689c1975dde">
                    <ee:message>
                        <ee:set-payload>output application/json --- {"error": "$(error.description)"}</ee:set-payload>
                        <ee:set-attributes>
                            {
                            statusCode: 404,
                            reasonPhrase : "$(error.description)"
                            }
                        </ee:set-attributes>
                    </ee:message>
                    <ee:variables>
                        <ee:set-variable variableName="httpStatus"><![CDATA[404]]></ee:set-variable>
                    </ee:variables>
                </ee:transform>
			</on-error-propagate>
            
			<on-error-propagate enableNotifications="true" logException="true" doc:name="On Error Propagate" doc:id="03ecdb66-d37b-4a4d-ab0a-a6d420db0181" type="HTTP:CONNECTIVITY">
				<logger level="INFO" doc:name="Logger" doc:id="4410b214-c931-4461-a742-a3ae8f087241" message="Error Occurred: statusCode: 500" />
				<ee:transform doc:name="Transform Message" doc:id="95d70fd9-adf1-4c73-82ab-1ab77cf7c3a2" >
					<ee:message >
						<ee:set-payload ><![CDATA[output application/json --- {"error": "$(error.description)"}]]></ee:set-payload>
						<ee:set-attributes ><![CDATA[
                            {
                            statusCode: 500,
                            reasonPhrase : "$(error.description)"
                            }
                        ]]></ee:set-attributes>
					</ee:message>
					<ee:variables>
                        <ee:set-variable variableName="httpStatus"><![CDATA[500]]></ee:set-variable>
                    </ee:variables>
				</ee:transform>
				<logger level="INFO" doc:name="Logger" doc:id="9ce6d6fb-7114-4752-9153-a0cf0ab70dac" message="Error Occurred: statusCode: 500" />
			</on-error-propagate>
        </error-handler>

    </flow>

    
</mule>
`

**Comment by manikmagar** [comment]
@harishsubramanian2577 I tried with your proxy and app can see errors being reported to Dynatrace when Validation 404 is raised.

![image](https://user-images.githubusercontent.com/877286/201425707-7bebd53d-12e1-4084-bdd6-1626b57e8566.png)


However compared with the one I see, your first screenshot actually looks like a trace for a successful transaction and not a resource_not_found from the validation module. Could you please confirm that there was an error involved in that transaction?

**Comment by harishsubramanian2577** [comment]
@manikmagar , the first screenshot is for the 2 404 transactions which looks as if they were successful and that is exactly the issue here. When i click on the transaction and check the trace, i do see the 404 response code though. I want those transactions reported as errors on the landing page where it lists all the transactions. Manik, I can explain better in a call. Please let me know your availability.

**Comment by manikmagar** [comment]
Hi @harishsubramanian2577, thanks for the clarification. I am able to reproduce this behavior with the XML config you shared earlier. The module is behaving correctly. If you notice the proxy XML you shared, the `http-request-config` element is declaring `http:success-status-code-validator` with all possible status codes returned by backend API - `0..599`. This will result in `http:request` element NOT really throwing a Mule ERROR. Since the event does contain any Mule Error, No error event will be captured by the module. You see the HTTP status code 404 in trace because the proxy is forwarding all attributes repose.

```
  <http:request config-ref="http-request-config" method="#[attributes.method]" path="#[attributes.maskedRequestPath]" doc:id="be1bb696-764d-400b-b1cd-726ca4245600">
        <http:headers><![CDATA[#[vars.proxyRequestHeaders]]]></http:headers>
        <http:uri-params><![CDATA[#[attributes.uriParams]]]></http:uri-params>
        <http:query-params><![CDATA[#[attributes.queryParams]]]></http:query-params>
        <http:response-validator>
            <http:success-status-code-validator values="0..599"/>
        </http:response-validator>
    </http:request>
```

If you remove that `http:response-validator` from that config and let the `http:request` throw 404 as an error (as normally it would do), you will see the error event on trace. Here is a trace I got with removing response-validator.

Hope this helps!

![image](https://user-images.githubusercontent.com/877286/201768699-36bba27f-7af4-4858-afd6-77b01a40ad83.png)

**Comment by harishsubramanian2577** [comment]
Hi @manikmagar , Thanks for pointing that out. After removing the http:response-validator, I see the expected behavior in Dynatrace for 404 errors thrown from back end api. 
But for "404:resource not found" errors which is caught by the rest:validator (invoking a non existent resource. For Ex:- hello_world1 instead of hello_world), i had to make the below code change in the getSpanStatus method of HttpProcessorCompoent.java

` private StatusCode getSpanStatus(boolean isServer, int statusCode) {
    StatusCode result;
    logger.info("isServer:" + String.valueOf(isServer));
    int maxStatus = isServer ? 500 : 400;

    logger.info("statusCode:" + Integer.toString(statusCode));
    logger.info("maxStatus:" + Integer.toString(maxStatus));

    if (statusCode >= 100 && statusCode < maxStatus) {
      if (statusCode >= 100 && statusCode < 400) {
        logger.info("result set to UNSET");
        result = StatusCode.UNSET;
      } else {
        logger.info("result set to ERROR");
        result = StatusCode.ERROR;
      }
    } else {
      logger.info("result set to ERROR");
      result = StatusCode.ERROR;
    }
    return result;
  }`

Basically, in this scenario, the statusCode was getting set to UNSET and hence the transaction was not getting marked as Error in dynatrace. So I had to add  if (statusCode >= 100 && statusCode < 400) condition and set StatusCode to UNSET else set to ERROR. After doing this change, the transaction for this scenario is getting marked as Error in dynatrace. Please let me know if this is ok.

**Comment by manikmagar** [comment]
@harishsubramanian2577 That `setSpanStatus` logic is as per the guidelines provided in optetelemetry [semantic conventions for HTTP Span Status](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/http/#status). We cannot break that compatibility by the change you proposed there. The span you are looking at is a SERVER span for Proxy's HTTP listener. It can only go in ERROR state for 500+ statuses. 

Instead of a POSTMAN, if a HTTP REQUEST makes a call to that proxy then the span representing that HTTP REQUEST consumer would properly get marked as ERROR state due to 404. You may test that.

I would NOT suggest making that code change since it breaks the OTEL guidelines.

**Comment by harishsubramanian2577** [comment]
Thanks @manikmagar , I see your point, but in my case the consumer would be some web app / some other system which would get instrumented using dynatrace one agent and mule will use otel agent. so the operations team may ask why the trace from Mule is not marked as error when the status code is 404. Possible explanation is since this is a 4xx series error which are basically Client errors due to an invalid request from client, mule or the server will not mark it as error.  Am I sounding right?

**Comment by manikmagar** [comment]
@harishsubramanian2577 this is correct and shall be true for any OTEL compliant HTTP Servers - 
> 4xx series error which are basically Client errors due to an invalid request from client, mule or the server will not mark it as error

If web app is compliant with OTEL, it should have marked the trace as ERRORED. If web app propagates the context, this module will use that context and Dynatrace should show you the errored trace together. 
 
I believe the question of 404 transactions is resolved now. I will close this issue now. Feel free to report any other issues you may see.
